include/utilities/realtime/dataman/DataMan.h

```diff
@@ -0,0 +1,468 @@
+        msg["offset"] = p_offset;
+        msg["timestep"] = p_timestep;
+        msg["tolerance"] = p_tolerance;
+        msg["priority"] = p_priority;
+        return put(p_data, msg);
+    }
+
+    virtual int put_begin(const void *p_data, json &p_jmsg)
+    {
+        check_shape(p_jmsg);
+        p_jmsg["profiling"] = m_profiling;
+        m_step_time = chrono::system_clock::now();
+        return 0;
+    }
+
+    virtual int put_end(const void *p_data, json &p_jmsg)
+    {
+        auto end = chrono::system_clock::now();
+        chrono::duration<double> duration = end - m_step_time;
+        m_profiling["total_manager_time"] =
+            m_profiling["total_manager_time"].get<double>() + duration.count();
+        m_profiling["total_mb"] =
+            m_profiling["total_mb"].get<size_t>() +
+            product(p_jmsg["varshape"], dsize(p_jmsg["dtype"])) / 1000000.0f;
+        duration = end - m_start_time;
+        m_profiling["total_workflow_time"] = duration.count();
+        m_profiling["workflow_mbs"] =
+            m_profiling["total_mb"].get<double>() /
+            m_profiling["total_workflow_time"].get<double>();
+        m_profiling["manager_mbs"] =
+            m_profiling["total_mb"].get<double>() /
+            m_profiling["total_manager_time"].get<double>();
+        if (p_jmsg["compressed_size"] != nullptr)
+            p_jmsg["putbytes"] = p_jmsg["compressed_size"].get<size_t>();
+        put_next(p_data, p_jmsg);
+        return 0;
+    }
+
+    virtual int put(const void *p_data, json p_jmsg) = 0;
+
+    int get(void *p_data, string p_doid, string p_var, string p_dtype,
+            vector<size_t> p_getshape, vector<size_t> p_varshape,
+            vector<size_t> p_offset, size_t p_timestep)
+    {
+        json msg;
+        msg["doid"] = p_doid;
+        msg["var"] = p_var;
+        msg["dtype"] = p_dtype;
+        msg["getshape"] = p_getshape;
+        msg["varshape"] = p_varshape;
+        msg["offset"] = p_offset;
+        msg["timestep"] = p_timestep;
+        return get(p_data, msg);
+    }
+
+    int get(void *p_data, string p_doid, string p_var, string &p_dtype,
+            vector<size_t> &p_varshape, size_t &p_timestep)
+    {
+        json msg;
+        msg["doid"] = p_doid;
+        msg["var"] = p_var;
+        return get(p_data, msg);
+    }
+
+    virtual int get(void *p_data, json &p_jmsg) = 0;
+    virtual int init(json p_jmsg) = 0;
+    virtual void flush() = 0;
+    virtual string name() = 0;
+    virtual string type() = 0;
+    void reg_callback(std::function<void(const void *, string, string, string,
+                                         vector<size_t>)>
+                          cb)
+    {
+        if (m_next.size() == 0)
+        {
+            m_callback = cb;
+        }
+        else
+        {
+            for (auto i : m_next)
+            {
+                i.second->reg_callback(cb);
+            }
+        }
+    }
+
+    void dump(const void *p_data, json p_jmsg)
```

Comment:
Rather than dumping to cout, take an arbitrary ostream that defaults to cout:
```
void dump(const void *p_data, json p_jmsg, std::ostream &out = std::cout)
```